core/dive.c

```diff
@@ -3020,30 +3020,38 @@ void taglist_cleanup(struct tag_entry **tag_list)
 	}
 }
 
-int taglist_get_tagstring(struct tag_entry *tag_list, char *buffer, int len)
+char* taglist_get_tagstring(struct tag_entry *tag_list)
 {
-	int i = 0;
-	struct tag_entry *tmp;
-	tmp = tag_list;
-	memset(buffer, 0, len);
+	char *buffer = NULL;
+	char *last_pos = NULL;
+	size_t buffer_size = 0;
+	struct tag_entry *tmp = tag_list;
+	/* Compute buffer size counting extra ", " separator(s) */
 	while (tmp != NULL) {
-		int newlength = strlen(tmp->tag->name);
-		if (i > 0)
-			newlength += 2;
-		if ((i + newlength) < len) {
-			if (i > 0) {
-				strcpy(buffer + i, ", ");
-				strcpy(buffer + i + 2, tmp->tag->name);
-			} else {
-				strcpy(buffer, tmp->tag->name);
-			}
-		} else {
-			return i;
+		if (buffer_size > 0) {
+			buffer_size += 2;
 		}
-		i += newlength;
+		buffer_size += strlen(tmp->tag->name);
 		tmp = tmp->next;
 	}
-	return i;
+	/* Allocate buffer and generate string with ", " separator(s) */
+	if (buffer_size > 0) {
```

Comment:
Please don't make it complicated for the caller (having to handle null). If there are no tags, return an empty string. Also makes the code here more straight-forward.

Reply:
i think it's OK to return `NULL` as this is also covered in the tests.

Reply:
I think it's a really bad interface. The next person does a very reasonable
```
char *s = taglist_get_tagstring(...);
snprintf(buf, sizeof(buf), "Tags: %s\n", s);
free(s);
```
And boom  - strange and undefined behavior.

This code makes it more complicated for the caller *and* the callee. What's the advantage?

Reply:
i guess we can do that. please return `""` instead of `NULL`.
but this will also affect the test.

Reply:
My idea was to remove both `if`-conditions - if there are not tags it would simply fall back to an empty string, no?

Reply:
let's see the next iteration that @dje29 comes up with.